### PR TITLE
[Typescript] add support for typescript es6 promises

### PR DESCRIPTION
**Pull request by wing328**
Ref: https://github.com/swagger-api/swagger-codegen/pull/2459

To use the option, pass `-DsupportsES6=true` or do it in config file.

cc @Bherila

### CODE REVIEW

**Comment by wing328** [comment]
also cc @Vrolijkx

File: samples/client/petstore/typescript-angular/tsd.json [hunk 1]
```diff
@@ -5,11 +5,20 @@
...
   "bundle": "typings/tsd.d.ts",
   "installed": {
-    "angularjs/angular.d.ts": {
+  	"angularjs/angular.d.ts": {
       "commit": "f6c8ca47193fb67947944a3170912672ac3e908e"
     },
     "jquery/jquery.d.ts": {
       "commit": "f6c8ca47193fb67947944a3170912672ac3e908e"
+    },
+    "request/request.d.ts": {
```
Comment (Vrolijkx) [inline samples/client/petstore/typescript-angular/tsd.json]: Why do we have a dependency on request in the angular client? Isn't request only for node?

**Comment by Vrolijkx** [comment]
Bluebird has all features that ES6 promises have. Can't we change all code to ES6 promises. And only when we need to support ES5  the following code is needed. 
```
{{^supportsES6}}
 import Promise = require('bluebird');
 {{/supportsES6}}
```
This will make the templates simpler because we avoid a lot of conditional blocks.

**Comment by wing328** [comment]
@Vrolijkx so you're suggesting making ES6 as the default? (currently the default is ES5)

**Comment by Vrolijkx** [comment]
@wing328 No is suggest to keep the default ES5 with bluebird. but wright all the code in with te ES6 promise syntax.
We would always write:
```typescript
return new Promise((reject, resolve) =>{
  request(requestOptions, (error, response, body) => {
    if (error) {
        reject(error);
    } else {
       //do some code
       resolve(THE_VALUE_TO_RETURN)
    }
})
```
instead of:
```typescript
let localVarDeferred = promise.defer<ReturnType>();
request(requestOptions, (error, response, body) => {
    if (error) {
        localVarDeferred.reject(error);
    } else {
       //do some code
       localVarDeferred.resolve(THE_VALUE_TO_RETURN)
    }
});
return localVarDeferred.promise
```
The first example also works in bluebird so why support two different code templates for these parts.

**Comment by wing328** [comment]
@Vrolijkx thanks for the explanation. If we switch to the ES6 promise syntax and when developers upgrade the Typescript SDK with the ES6 promise syntax, do you need to make any code change or the new ES6 promise syntax is backward compatible? (this is not to say we cannot make breaking change but I need to understand the impact to the developers/users)

**Comment by Vrolijkx** [comment]
The new syntax is backward compatible. For a consumer of an api no changes are needed.

File: samples/client/petstore/typescript-angular/tsd.json (hunk 1 above)
Comment (Vrolijkx) [inline samples/client/petstore/typescript-angular/tsd.json]: Maybe the sample generation script updates the wrong values.

**Comment by bherila** [comment]
I seem to recall that Bluebird deprecated the use of Promise.defer anyways. So that's another reason to switch over to the ES6 syntax.

**Comment by wing328** [comment]
@Vrolijkx thanks for the confirmation. I've merged this PR into master for the time being. Please feel free to submit a PR with the suggested change (so that the credit goes back to your Github account). 

If you need help, just let me know.

**Pull request merged**